Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -760,128 +748,147 @@ AzFramework::RenderGeometry::RayResult TerrainSystem::GetClosestIntersection(
-    const AZStd::span<const AZ::Vector3>& inPositions,
-    AzFramework::Terrain::SurfacePointListFillCallback perPositionCallback,
-    Sampler sampler,
-    AZStd::shared_ptr<ProcessAsyncParams> params) const
-{
-    return ProcessFromListAsync(AZStd::bind(&TerrainSystem::ProcessSurfacePointsFromList, this, AZStd::placeholders::_1, std::placeholders::_2, std::placeholders::_3),
-        inPositions, perPositionCallback, sampler, params);
-}
+    auto numSamplesX = queryRegion.m_numPointsX;
+    auto numSamplesY = queryRegion.m_numPointsY;
+    const int64_t numPositionsToProcess = numSamplesX * numSamplesY;
 
-AZStd::shared_ptr<AzFramework::Terrain::TerrainDataRequests::TerrainJobContext> TerrainSystem::ProcessHeightsFromListOfVector2Async(
-    const AZStd::span<const AZ::Vector2>& inPositions,
-    AzFramework::Terrain::SurfacePointListFillCallback perPositionCallback,
-    Sampler sampler,
-    AZStd::shared_ptr<ProcessAsyncParams> params) const
-{
-    return ProcessFromListAsync(AZStd::bind(&TerrainSystem::ProcessHeightsFromListOfVector2, this, AZStd::placeholders::_1, std::placeholders::_2, std::placeholders::_3),
-        inPositions, perPositionCallback, sampler, params);
-}
+    if (numPositionsToProcess == 0)
+    {
+        AZ_Warning("TerrainSystem", false, "No positions to process.");
+        return nullptr;
+    }
 
-AZStd::shared_ptr<AzFramework::Terrain::TerrainDataRequests::TerrainJobContext> TerrainSystem::ProcessNormalsFromListOfVector2Async(
-    const AZStd::span<const AZ::Vector2>& inPositions,
-    AzFramework::Terrain::SurfacePointListFillCallback perPositionCallback,
-    Sampler sampler,
-    AZStd::shared_ptr<ProcessAsyncParams> params) const
-{
-    return ProcessFromListAsync(AZStd::bind(&TerrainSystem::ProcessNormalsFromListOfVector2, this, AZStd::placeholders::_1, std::placeholders::_2, std::placeholders::_3),
-        inPositions, perPositionCallback, sampler, params);
-}
+    // Determine the maximum number of jobs, and the minimum number of positions that should be processed per job.
+    const int32_t numJobsMax = CalculateMaxJobs(params);
+    const int32_t minPositionsPerJob = params && (params->m_minPositionsPerJob > 0)
+        ? params->m_minPositionsPerJob
+        : AzFramework::Terrain::QueryAsyncParams::MinPositionsPerJobDefault;
+
+    // Calculate the best subdivision of the region along both the X and Y axes to use as close to the maximum number of jobs
+    // as possible while also keeping all the regions effectively the same size.
+    int32_t xJobs, yJobs;
+    SubdivideRegionForJobs(
+        aznumeric_cast<int32_t>(numSamplesX), aznumeric_cast<int32_t>(numSamplesY), numJobsMax, minPositionsPerJob, xJobs, yJobs);
+
+    // The number of jobs returned might be less than the total requested maximum number of jobs, so recalculate it here
+    AZ_Assert((xJobs * yJobs) <= numJobsMax, "The region was subdivided into too many jobs: %d x %d vs %d max", xJobs, yJobs, numJobsMax);
```

Comment:
Since `SubdivideRegionForJobs()` is privately owned by TerrainSystem, it seems weird to have an assert message checking its output here. I could see maybe having an assert inside `SubdivideRegionForJobs()` to make sure if a programmer makes changes to the code they don't mess this up.